aspnetcore/blazor/webassembly-native-dependencies.md

```diff
@@ -0,0 +1,159 @@
+
+   ```razor
+   @using System.Runtime.InteropServices
+
+   <p>@fact(3)</p>
+
+   @code {
+       [DllImport("Test")]
+       static extern int fact(int n);
+   }
+   ```
+
+When you build the app with the .NET WebAssembly build tools installed, the native C code is compiled and linked into the .NET WebAssembly runtime (`dotnet.wasm`). Compiling and linking may take a few minutes. After the app is built, run the app to see the rendered factorial value.
+
+> [!NOTE]
+> During the 6.0 preview release period, you may receive a build error on subsequent builds saying that the output assembly is being used by another process. This is a known issue that will be addressed for the .NET 6 general release. To workaround the issue, rebuild the project a second time.
+
+## C++ managed method callbacks
+
+Label managed methods that are passed to C++ with the `[UnmanagedCallersOnly]` attribute.
+
+The method marked with the `[UnmanagedCallersOnly]` attribute must be `static`. To call an instance method in a Razor component, pass a `GCHandle` for the instance to C++ and then pass it back to native. Alternatively, use some other method to identify the instance of the component.
+
+The method marked with `[DllImport]` must use a C# 9.0 function pointer rather than a delegate type for the callback argument.
+
+> [!NOTE]
+> For C# function pointer types in `[DllImport]` methods, use `IntPtr` in the method signature on the managed side instead of `delegate *unmanaged<int, void>`. For more information, see [[WASM] callback from native code to .NET: Parsing function pointer types in signatures is not supported (dotnet/runtime #56145)](https://github.com/dotnet/runtime/issues/56145).
+
+## Use libraries
+
+NuGet packages can contain native dependencies for use on WebAssembly. These libraries and their native functionality are then available to any Blazor WebAssembly app. The files for the native dependencies should be built for WebAssembly and packaged in the `browser-wasm` [architecture-specific folder](/nuget/create-packages/supporting-multiple-target-frameworks#architecture-specific-folders). WebAssembly-specific dependencies aren't referenced automatically and must be referenced manually as `NativeFileReference`s. Package authors can choose to add the native references by including a `.props` file in the package with the references.
+
+[SkiaSharp](https://github.com/mono/SkiaSharp) is a cross-platform 2D graphics library for .NET based on the native [Skia graphics library](https://skia.org/), and it now has preview support for Blazor WebAssembly.
+
+> [!WARNING]
+> [SkiaSharp](https://github.com/mono/SkiaSharp) and the [Skia graphics library](https://skia.org/) aren't owned or maintained by Microsoft. The demonstration example in this section is unsupported by Microsoft and shouldn't be used in production.
```

Comment:
> shouldn't be used in production.

That's a strong statement. I think what we mean to say is something like "do this if you want but don't blame us if there's a problem", whereas it sounds like "you should not do this because it's a bad thing to do". We're not actually trying to say there's something bad about the following code, right? We just mean to communicate that we're not officially promising anything about it.

Would it be enough simply to say "... is unsupported by Microsoft"? I think that communicates the intent.

Reply:
I yield to @danroth27. We have a few spots where we say that preview/experimental packages are not supported for production use. For example ...

> `Microsoft.AspNetCore.ProtectedBrowserStorage` is an unsupported, experimental package that isn't intended for production use.

> The `Microsoft.AspNetCore.Components.WebAssembly.MultipartBundle` NuGet package and the `CustomPackagedApp` sample app are unsupported, experimental demonstration resources not intended for production use.

... and IIRC, we did something similar for `AspNetCore.Diagnostics.HealthChecks` while it was in preview. My intention is to remove that part of the remark when SkiaSharp goes off of preview.

I'll 👂 @danroth27 for guidance on these examples and update them all on this PR.

Reply:
I'm asking Matthew if he has an ETA on a RTM/GA release at [Is there an ETA on RTM/GA for SkiaSharp.Views.Blazor? (mono/SkiaSharp #1862)](https://github.com/mono/SkiaSharp/issues/1862). I'll hold this PR for a sec to see if he says that the release is imminent. If so, I'll strike the warning about production use and leave our usual 'not supported by MS' remark.

Reply:
Users are free to use whatever features/code they want in production, as long as they've tested that they satisfy their needs. Pretty much all of .NET, stable or not, is provided "as is" as stated in our various licenses. So, we really only need to clarify the level of support and intent to release a stable version.

Suggested language: "Experimental and preview features are provided for the purpose of collecting feedback and are not supported for production use." 

The difference between experimental and preview features is that preview features are expect to eventual ship a stable release, while experimental features may not. Suggested language: "Experimental features are provided for the purpose of exploring feature viability and may not ship a future stable version".